File: ipc-codegen/echo_example/cpp/src/echo_client.cpp
// Echo IPC client (C++) — uses the generated EchoIpcClient.
// Usage: echo_client --socket /tmp/echo.sock

#include "generated/echo_ipc_client.hpp"

#include <array>
#include <iostream>
#include <string_view>

// Explicit check (not assert): verification must survive NDEBUG builds.
#define CHECK(cond, label)                                                     \
  do {                                                                         \
    if (!(cond)) {                                                             \
      std::cerr << "echo_client(cpp): " << (label) << " FAIL\n";               \
      return 1;                                                                \
    }                                                                          \
  } while (0)

namespace {
echo::Fr test_hash(uint8_t base) {
  echo::Fr hash{};
  for (size_t i = 0; i < hash.size(); ++i) {
    hash[i] = static_cast<uint8_t>(base + i);
  }
  return hash;
}
} // namespace

int main(int argc, char **argv) {
  const char *socket_path = nullptr;
  for (int i = 1; i < argc - 1; i++) {
    if (std::string_view(argv[i]) == "--socket")
      socket_path = argv[i + 1];
  }
  if (!socket_path) {
    std::cerr << "Usage: echo_client --socket <path>\n";
    return 1;
  }

  echo::EchoIpcClient client(socket_path);

  {
    auto resp = client.bytes({.data = {0xDE, 0xAD, 0xBE, 0xEF, 0x42}});
    CHECK((resp.data == std::vector<uint8_t>{0xDE, 0xAD, 0xBE, 0xEF, 0x42}),
          "EchoBytes");
    std::cerr << "echo_client(cpp): EchoBytes OK\n";
  }

  {
    auto resp =
        client.fields({.a = 42, .b = 999999, .name = "hello wire compat"});
    CHECK(resp.a == 42 && resp.b == 999999 && resp.name == "hello wire compat",
          "EchoFields");
    std::cerr << "echo_client(cpp): EchoFields OK\n";
  }

  {
    auto resp =
        client.nested({.inner = {.values = {{1, 2, 3}, {4, 5}}, .flag = true}});
    CHECK((resp.inner.values ==
           std::vector<std::vector<uint8_t>>{{1, 2, 3}, {4, 5}}),
          "EchoNested values");
    CHECK(resp.inner.flag == true, "EchoNested flag");
    std::cerr << "echo_client(cpp): EchoNested OK\n";
  }

  {
    auto hash = test_hash(0x10);
    auto second = test_hash(0x40);
    auto resp = client.aliases({.treeId = 7,
                                .hash = hash,
                                .maybeHash = second,
                                .hashes = {hash, second}});
    CHECK(resp.treeId == 7, "EchoAliases treeId");
    CHECK(resp.hash == hash, "EchoAliases hash");
    CHECK(resp.maybeHash == second, "EchoAliases maybeHash");
    CHECK((resp.hashes == std::vector<echo::Fr>{hash, second}),
          "EchoAliases hashes");
    std::cerr << "echo_client(cpp): EchoAliases OK\n";
  }

  // Optional-absent over live IPC.
  {
    auto hash = test_hash(0x10);
    auto resp = client.aliases({.treeId = 7,
                                .hash = hash,
                                .maybeHash = std::nullopt,
                                .hashes = {hash}});
    CHECK(!resp.maybeHash.has_value(), "EchoAliases none");
    std::cerr << "echo_client(cpp): EchoAliases none OK\n";
  }

  // uint64 wire encoding above 2^32 over live IPC.
  {
    const uint64_t big = (1ULL << 53) - 1;
    auto resp = client.fields({.a = 42, .b = big, .name = "big"});
    CHECK(resp.b == big, "EchoFields u64");
    std::cerr << "echo_client(cpp): EchoFields u64 OK\n";
  }

  // Optional bytes Some/None and fixed [bytes; 2].
  {
    auto resp = client.blobs({.maybeData = std::vector<uint8_t>{0xAA, 0xBB},
                              .parts = {{{1, 2, 3}, {4}}}});
    CHECK((resp.maybeData == std::vector<uint8_t>{0xAA, 0xBB}),
          "EchoBlobs maybeData");
    CHECK((resp.parts == std::array<std::vector<uint8_t>, 2>{{{1, 2, 3}, {4}}}),
          "EchoBlobs parts");
    auto resp_none =
        client.blobs({.maybeData = std::nullopt, .parts = {{{}, {9}}}});
    CHECK(!resp_none.maybeData.has_value(), "EchoBlobs none");
    std::cerr << "echo_client(cpp): EchoBlobs OK\n";
  }

  // Server error surfaces with its message.
  {
    bool threw = false;
    std::string message;
    try {
      client.fail({.message = "deliberate failure"});
    } catch (const std::exception &e) {
      threw = true;
      message = e.what();
    }
    CHECK(threw, "EchoFail threw");
    CHECK(message.find("deliberate failure") != std::string::npos,
          "EchoFail message");
    std::cerr << "echo_client(cpp): EchoFail OK\n";
  }

  std::cerr << "echo_client(cpp): all tests passed\n";
  return 0;
}

File: ipc-codegen/echo_example/ts/src/echo_client.ts
/**
 * Echo IPC client (TypeScript) — uses the GENERATED AsyncApi client over the
 * @aztec/ipc-runtime transport. Defaults to UDS; pass `--transport shm` to
 * drive the bundled NAPI SHM client (`createNapiShmAsyncClient`) instead.
 * Path suffix follows the same convention as ipc::make_client on the C++
 * side: `.sock` for UDS, `.shm` for MPSC-SHM rings.
 *
 * Usage: npx tsx echo_client.ts --socket /tmp/echo.sock [--transport uds|shm]
 * Exits 0 on success, 1 on failure.
 */
import {
  createNapiShmAsyncClient,
  UdsIpcClient,
  type IpcClientAsync,
} from "@aztec/ipc-runtime";
import { AsyncApi } from "./generated/async.js";

const args = process.argv.slice(2);
const socketIdx = args.indexOf("--socket");
const socketPath = socketIdx >= 0 ? args[socketIdx + 1] : undefined;
if (!socketPath) {
  console.error("Usage: echo_client.ts --socket <path> [--transport uds|shm]");
  process.exit(1);
}

function testHash(base: number): Uint8Array {
  return Uint8Array.from({ length: 32 }, (_v, i) => base + i);
}
const transportIdx = args.indexOf("--transport");
const transport = transportIdx >= 0 ? args[transportIdx + 1] : "uds";
if (transport !== "uds" && transport !== "shm") {
  console.error(`Unknown --transport '${transport}' (expected uds|shm)`);
  process.exit(1);
}

function assertEqual(actual: unknown, expected: unknown, label: string) {
  const a = JSON.stringify(actual);
  const e = JSON.stringify(expected);
  if (a !== e) throw new Error(`${label}: expected ${e}, got ${a}`);
}

function assertBytes(actual: Uint8Array, expected: Uint8Array, label: string) {
  assertEqual(
    Buffer.from(actual).toString("hex"),
    Buffer.from(expected).toString("hex"),
    label,
  );
}

async function run() {
  // SHM clients identify the shared-memory base name without the `.shm`
  // suffix — match ipc::make_client's behaviour on the C++ side.
  const client: IpcClientAsync =
    transport === "shm"
      ? createNapiShmAsyncClient(socketPath!.replace(/\.shm$/, ""))
      : await UdsIpcClient.connect(socketPath!);
  const api = new AsyncApi(client);

  // Test 1: EchoBytes
  const testData = Uint8Array.from([0xde, 0xad, 0xbe, 0xef, 0x42]);
  const resp1 = await api.echoBytes({ data: testData });
  assertBytes(resp1.data, testData, "EchoBytes data");
  console.error("echo_client(ts): EchoBytes OK");

  // Test 2: EchoFields
  const resp2 = await api.echoFields({
    a: 42,
    b: 999999,
    name: "hello wire compat",
  });
  assertEqual(resp2.a, 42, "EchoFields a");
  assertEqual(resp2.b, 999999, "EchoFields b");
  assertEqual(resp2.name, "hello wire compat", "EchoFields name field");
  console.error("echo_client(ts): EchoFields OK");

  // Test 3: EchoNested
  const inner = {
    values: [Uint8Array.from([1, 2, 3]), Uint8Array.from([4, 5])],
    flag: true,
  };
  const resp3 = await api.echoNested({ inner });
  assertEqual(resp3.inner.flag, true, "EchoNested flag");
  assertEqual(resp3.inner.values.length, 2, "EchoNested values length");
  assertBytes(resp3.inner.values[0]!, inner.values[0]!, "EchoNested values[0]");
  console.error("echo_client(ts): EchoNested OK");

  // Test 4: EchoAliases
  const hash = testHash(0x10);
  const second = testHash(0x40);
  const resp4 = await api.echoAliases({
    treeId: 7,
    hash,
    maybeHash: second,
    hashes: [hash, second],
  });
  assertEqual(resp4.treeId, 7, "EchoAliases treeId");
  assertBytes(resp4.hash, hash, "EchoAliases hash");
  assertBytes(resp4.maybeHash!, second, "EchoAliases maybeHash");
  assertEqual(resp4.hashes.length, 2, "EchoAliases hashes length");
  assertBytes(resp4.hashes[0]!, hash, "EchoAliases hashes[0]");
  assertBytes(resp4.hashes[1]!, second, "EchoAliases hashes[1]");
  console.error("echo_client(ts): EchoAliases OK");

  // Test 5: EchoAliases with maybeHash absent (optional over live IPC)
  const resp5 = await api.echoAliases({
    treeId: 7,
    hash,
    maybeHash: null,
    hashes: [hash],
  });
  assertEqual(resp5.maybeHash, null, "EchoAliases maybeHash none");
  console.error("echo_client(ts): EchoAliases none OK");

  // Test 6: EchoFields with b > 2^32 (uint64 wire encoding over live IPC)
  const big = Number.MAX_SAFE_INTEGER;
  const resp6 = await api.echoFields({ a: 42, b: big, name: "big" });
  assertEqual(resp6.b, big, "EchoFields u64");
  // Values past 2^53 must throw client-side rather than silently lose precision.
  let threw = false;
  try {
    await api.echoFields({ a: 42, b: 2 ** 60, name: "too big" });
  } catch {
    threw = true;
  }
  assertEqual(threw, true, "EchoFields u64 guard");
  console.error("echo_client(ts): EchoFields u64 OK");

  // Test 7: EchoBlobs — optional bytes Some/None and fixed [bytes; 2]
  const resp7 = await api.echoBlobs({
    maybeData: Uint8Array.from([0xaa, 0xbb]),
    parts: [Uint8Array.from([1, 2, 3]), Uint8Array.from([4])],
  });
  assertBytes(
    resp7.maybeData!,
    Uint8Array.from([0xaa, 0xbb]),
    "EchoBlobs maybeData",
  );
  assertBytes(
    resp7.parts[0]!,
    Uint8Array.from([1, 2, 3]),
    "EchoBlobs parts[0]",
  );
  assertBytes(resp7.parts[1]!, Uint8Array.from([4]), "EchoBlobs parts[1]");
  const resp7b = await api.echoBlobs({
    maybeData: null,
    parts: [Uint8Array.from([]), Uint8Array.from([9])],
  });
  assertEqual(resp7b.maybeData, null, "EchoBlobs maybeData none");
  console.error("echo_client(ts): EchoBlobs OK");

  // Test 8: EchoFail — server error surfaces with its message
  let failMessage = "";
  try {
    await api.echoFail({ message: "deliberate failure" });
  } catch (e: any) {
    failMessage = e.message;
  }
  if (!failMessage.includes("deliberate failure")) {
    throw new Error(`EchoFail: expected error message, got '${failMessage}'`);
  }
  console.error("echo_client(ts): EchoFail OK");

  await api.destroy();
  console.error("echo_client(ts): all tests passed");
}

run().catch((e) => {
  console.error(`echo_client(ts): FAILED: ${e.message}`);
  process.exit(1);
});
